odl/space/npy_tensors.py

```diff
@@ -2019,27 +2113,112 @@ def _pnorm_diagweight(x, p, w):
         return np.sum(xp) ** (1 / p)
 
 
-def _inner_default(x1, x2):
+def _inner_impl(x1, x2):
     """Default Euclidean inner product implementation."""
     # Ravel both in the same order
-    order = 'F' if all(a.data.flags.f_contiguous for a in (x1, x2)) else 'C'
+    order = 'F' if all(a.flags.f_contiguous for a in (x1, x2)) else 'C'
 
     if is_real_dtype(x1.dtype):
         if x1.size > THRESHOLD_MEDIUM:
             # This is as fast as BLAS dotc
             return np.tensordot(x1, x2, [range(x1.ndim)] * 2)
         else:
             # Several times faster for small arrays
-            return np.dot(x1.data.ravel(order),
-                          x2.data.ravel(order))
+            return np.dot(x1.ravel(order), x2.ravel(order))
     else:
         # x2 as first argument because we want linearity in x1
-        return np.vdot(x2.data.ravel(order),
-                       x1.data.ravel(order))
+        return np.vdot(x2.ravel(order), x1.ravel(order))
+
 
+# --- Weightings --- #
 
-# TODO: implement intermediate weighting schemes with arrays that are
-# broadcast, i.e. between scalar and full-blown in dimensionality?
+
+def slice_weighting(weighting, space_shape, indices):
+    """Return a weighting for a space after indexing.
```

Comment:
I had this in mind as well. That would mean putting `__getitem__` and `byaxis` on the weighting classes, and to raise an exception where it doesn't make sense. I'd be okay with that, it would also be more extensible.

Reply:
I feel it should be done that way, much more future proof

Reply:
Looking at the code again regarding this one, I'm having doubts about `__getitem__`. The issue is that in quite a few cases, the result after slicing is not the same class as before. That would seem okay for such simple things like (array of 1 element -> scalar) transformations, but here we make much bigger changes, like `ArrayWeighting` -> `None`. This would feel wrong in a `__getitem__`.